### PR TITLE
Further threading optimizations + caching

### PR DESCRIPTION
**Pull request by naemono**
Further threading/concurrent optimizations
Using disk caching for extremely large sensu installations

### CODE REVIEW

**Review by alex-leonhardt — COMMENTED** [review]
Never used concurrent.futures so this will be a nice start 👍  ... I've a few comments though re the caching ..

File: README.md [hunk 1]
```diff
@@ -128,6 +128,7 @@ app:
   # This is a python requests layer timeout, as by default, it does not timeout
   requests_timeout: 10
   logging_level: info
+  cache_expire_time: 60
```
Comment (alex-leonhardt) [inline README.md]: i think i'd prefer a very low default cache_expire_time setting (and would suggest so in the doc too); most installations are probably not _very_ big or might not even want to enable caching.

File: sensugrid.py [hunk 2]
```diff
@@ -17,17 +21,18 @@
...
 
-from multiprocessing.dummy import Pool as ThreadPool
-# https://stackoverflow.com/questions/2846653/how-to-use-threading-in-python
-
 import json
 import logging
 import logging.config
 
+from diskcache import Cache
+cache = Cache('/tmp/sensugrid')
```
Comment (alex-leonhardt) [inline sensugrid.py]: this is probably better off somewhere in /var/tmp/ as /tmp/ gets cleaned up periodically - that might cause some confusion to the app when the file is suddenly gone

File: sensugrid.py [hunk 3]
```diff
@@ -95,21 +101,48 @@ def _cmp(x, y):
...
-        aggregated = pool.map(get_agg_data, dcs)
-    except Exception as e:
-        print("Exception: ", e)
-    finally:
-        pool.close()
-    return render_template('data.html', dcs=dcs, data=aggregated, filter_data=get_filter_data(dcs, timeout), appcfg=appcfg)
+    futures = list()
+    filters = list()
+    agg_data_executor = concurrent.futures.ThreadPoolExecutor(len(dcs))
+    [futures.append(agg_data_executor.submit(get_agg_data, dc)) for dc in dcs]
```
Comment (alex-leonhardt) [inline sensugrid.py]: i agree that list comprehension seems more elegant, but it's less readable; can this be the "normal"
```
for dc in dcs:
    futures.append......
```
?

**Comment by naemono** [comment]
@alex-leonhardt all done.

**Comment by alex-leonhardt** [comment]
Thanks, had a quick look tonight and there's something odd going on with those `Filter by` drop-downs again .. 

screen shot 2018-01-26 at 07 25 54

**Comment by alex-leonhardt** [comment]
noticed another issue - the DCs keep jumping out of order .. 
screen shot 2018-02-02 at 07 04 08

**Comment by alex-leonhardt** [comment]
@naemono merged 👍  and managed to find a relatively simple fix for both issues above

**Comment by naemono** [comment]
Thanks @alex-leonhardt.  I hadn't had a chance to get back to this yet.  I appreciate the fix.

**Pull request merged**